MDANSE_GUI/Src/MDANSE_GUI/Tabs/Models/TrajectoryModel.py

```diff
@@ -0,0 +1,156 @@
+#
+
+from qtpy.QtCore import (
+    QDeadlineTimer,
+    QModelIndex,
+    QMutex,
+    QObject,
+    Qt,
+    QThread,
+    Signal,
+    Slot,
+)
+from qtpy.QtGui import QStandardItem, QStandardItemModel
+
+from MDANSE.MLogging import LOG
+from MDANSE.MolecularDynamics.Trajectory import Trajectory
+
+
+class LoaderThread(QThread):
+    results = Signal(object)
+
+    def __init__(
+        self,
+        parent,
+        filename: str,
+        index: int,
+        *,
+        display_item: QStandardItem | None = None,
+        target_label: str | None = None,
+    ):
+        super().__init__(parent)
+        self._filename = filename
+        self._target_index = index
+        self._display_item = display_item
+        self._target_label = target_label
+
+    def run(self):
+        trajectory = Trajectory(self._filename)
+        self.results.emit((trajectory, self._target_index))
+        if self._display_item is not None and self._target_label is not None:
+            self._display_item.setData(
+                self._target_label, role=Qt.ItemDataRole.DisplayRole
+            )
+
+
+class TrajectoryModel(QStandardItemModel):
+    """Like GeneralModel, but should implement trajectory
+    loading in the background."""
+
+    error = Signal(str)
+    all_elements = Signal(object)
+    finished_loading = Signal(int)
+    free_name = Signal(str)
+
+    def __init__(self, parent: QObject = None):
+        super().__init__(parent=parent)
+        self.mutex = QMutex()
+        self._node_numbers = []
+        self._trajectory_paths = {}
+        self._trajectory_instances = {}
+        self._loading_threads = {}
+        self._next_number = 0
+
+    @Slot(tuple)
+    def append_object(self, input: tuple):
+        full_name, label = input
+        self._node_numbers.append(self._next_number)
+        self._trajectory_paths[self._next_number] = full_name
+        self._trajectory_instances[self._next_number] = None
+        retval = int(self._next_number)
+        self._next_number += 1
+        item = QStandardItem(f"Loading {label}...")
+        item.setData(retval)
+        self.appendRow(item)
+        self.launch_loader(full_name, retval, display_item=item, target_label=label)
+        return retval
+
+    def launch_loader(
+        self,
+        filename: str,
+        index: int,
+        display_item: QStandardItem | None = None,
+        target_label: str | None = None,
+    ):
+        thread = LoaderThread(
+            None, filename, index, display_item=display_item, target_label=target_label
+        )
+        thread.results.connect(self.accept_results)
+        self._loading_threads[index] = thread
+        thread.start()
+
+    def get_trajectory(self, index: int):
+        result = None
+        if index not in self._loading_threads:
+            LOG.info("Requesting a missing trajectory with index %s", index)
+            return result
+        elif index not in self._trajectory_instances:
```

Comment:
```suggestion
        if index not in self._trajectory_instances:
```
Because the other is a `return`